File: clients/java/src/test/java/com/quidnug/client/VectorsTest.java
package com.quidnug.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Java SDK consumer for the v1.0 cross-SDK test vectors at
 * {@code docs/test-vectors/v1.0/}.
 *
 * Asserts the five conformance properties against the Java SDK's
 * public API ({@link Quid#sign}, {@link Quid#verify},
 * {@link Quid#fromPublicHex}, {@link Quid#fromPrivateScalarHex}).
 *
 * Run with: {@code mvn test -Dtest=VectorsTest}.
 */
class VectorsTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static Path vectorsRoot() {
        // Maven test runs from clients/java; vectors live at
        // <repo>/docs/test-vectors/v1.0.
        Path start = Paths.get(".").toAbsolutePath().normalize();
        Path p = start.resolve("../../docs/test-vectors/v1.0").normalize();
        if (!Files.isDirectory(p)) {
            throw new RuntimeException("vectors root not found at " + p);
        }
        return p;
    }

    private static Map<String, JsonNode> loadKeys() throws IOException {
        Map<String, JsonNode> out = new HashMap<>();
        File[] files = vectorsRoot().resolve("test-keys").toFile().listFiles(
                (d, n) -> n.endsWith(".json"));
        if (files == null) throw new RuntimeException("no keys dir");
        for (File f : files) {
            JsonNode k = MAPPER.readTree(f);
            out.put(k.get("name").asText(), k);
        }
        return out;
    }

    private static JsonNode loadVectorFile(String name) throws IOException {
        return MAPPER.readTree(vectorsRoot().resolve(name).toFile());
    }

    /**
     * Execute all conformance properties for a single case. The
     * caller supplies the canonical signable bytes + derived ID
     * built via whatever tx-type-specific builder applies.
     */
    private static void runCase(
            JsonNode caseNode, JsonNode key,
            byte[] signable, String derivedId) throws Exception {

        JsonNode expected = caseNode.get("expected");
        String name = caseNode.get("name").asText();

        // Property 1: sha256 of canonical bytes matches.
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        String gotSha = Hex.encode(md.digest(signable));
        assertEquals(
                expected.get("sha256_of_canonical_hex").asText(),
                gotSha,
                name + ": SHA-256 mismatch");

        // Property 2: canonical hex + utf8 match the vector.
        assertEquals(
                expected.get("canonical_signable_bytes_hex").asText(),
                Hex.encode(signable),
                name + ": canonical hex diverges");
        assertEquals(
                expected.get("canonical_signable_bytes_utf8").asText(),
                new String(signable, StandardCharsets.UTF_8),
                name + ": canonical utf8 diverges");

        // Property 3: derived ID matches.
        assertEquals(
                expected.get("expected_id").asText(),
                derivedId,
                name + ": ID derivation mismatch");

        // Property 4: reference signature verifies via Quid.verify.
        Quid qRO = Quid.fromPublicHex(key.get("public_key_sec1_hex").asText());
        assertTrue(
                qRO.verify(signable, expected.get("reference_signature_hex").asText()),
                name + ": SDK Verify rejected reference signature");
        assertEquals(
                64,
                expected.get("signature_length_bytes").asInt(),
                name + ": signature_length_bytes != 64");

        // Property 5: tampered signature rejects.
        byte[] tampered = Hex.decode(expected.get("reference_signature_hex").asText());
        tampered[5] ^= 0x01;
        assertFalse(
                qRO.verify(signable, Hex.encode(tampered)),
                name + ": tampered signature accepted");

        // Property 6: independent SDK sign-verify round-trip.
        Quid qSign = Quid.fromPrivateScalarHex(key.get("private_scalar_hex").asText());
        String sdkSig = qSign.sign(signable);
        byte[] sdkSigBytes = Hex.decode(sdkSig);
        assertEquals(64, sdkSigBytes.length, name + ": SDK signature not 64 bytes");
        assertTrue(qSign.verify(signable, sdkSig),
                name + ": SDK sign-verify round-trip failed");

        // Property 7: quid_id derivation matches.
        byte[] pubBytes = Hex.decode(key.get("public_key_sec1_hex").asText());
        String derivedQuid = Hex.encode(java.util.Arrays.copyOf(
                md.digest(pubBytes), 8));
        assertEquals(key.get("quid_id").asText(), derivedQuid,
                name + ": quid_id derivation mismatch");
    }

    // -----------------------------------------------------------
    // Per-type signable-bytes builders (Go-struct-order JSON)
    // -----------------------------------------------------------
    //
    // Jackson's ObjectNode preserves insertion order as of
    // jackson-databind 2.0+, so the same pattern as pkg/client
    // (explicit field-by-field construction) produces the canonical
    // bytes we want.

    private static byte[] trustSignable(JsonNode inp) {
        var node = MAPPER.createObjectNode();
        node.put("id", inp.get("id").asText());
        node.put("type", "TRUST");
        node.put("trustDomain", inp.get("trustDomain").asText());
        node.put("timestamp", inp.get("timestamp").asLong());
        node.put("signature", "");
        node.put("publicKey", inp.get("publicKey").asText());
        node.put("truster", inp.get("truster").asText());
        node.put("trustee", inp.get("trustee").asText());
        // Float: use Go's convention via serializeGoCompatFloat.
        putGoCompatFloat(node, "trustLevel", inp.get("trustLevel").asDouble());
        node.put("nonce", inp.get("nonce").asLong());
        if (inp.hasNonNull("description") && !inp.get("description").asText().isEmpty()) {
            node.put("description", inp.get("description").asText());
        }
        if (inp.hasNonNull("validUntil") && inp.get("validUntil").asLong() != 0) {
            node.put("validUntil", inp.get("validUntil").asLong());
        }
        return nodeToBytes(node);
    }

    private static String trustId(JsonNode inp) {
        try {
            var seed = MAPPER.createObjectNode();
            seed.put("Truster", inp.get("truster").asText());
            seed.put("Trustee", inp.get("trustee").asText());
            putGoCompatFloat(seed, "TrustLevel", inp.get("trustLevel").asDouble());
            seed.put("TrustDomain", inp.get("trustDomain").asText());
            seed.put("Timestamp", inp.get("timestamp").asLong());
            byte[] bytes = MAPPER.writeValueAsBytes(seed);
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return Hex.encode(md.digest(bytes));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] identitySignable(JsonNode inp) {
        var node = MAPPER.createObjectNode();
        node.put("id", inp.get("id").asText());
        node.put("type", "IDENTITY");
        node.put("trustDomain", inp.get("trustDomain").asText());
        node.put("timestamp", inp.get("timestamp").asLong());
        node.put("signature", "");
        node.put("publicKey", inp.get("publicKey").asText());
        node.put("quidId", inp.get("quidId").asText());
        node.put("name", inp.get("name").asText());
        if (inp.hasNonNull("description") && !inp.get("description").asText().isEmpty()) {
            node.put("description", inp.get("description").asText());
        }
        if (inp.hasNonNull("attributes") && inp.get("attributes").size() > 0) {
            node.set("attributes", inp.get("attributes"));
        }
        node.put("creator", inp.get("creator").asText());
        node.put("updateNonce", inp.get("updateNonce").asLong());
        if (inp.hasNonNull("homeDomain") && !inp.get("homeDomain").asText().isEmpty()) {
            node.put("homeDomain", inp.get("homeDomain").asText());
        }
        return nodeToBytes(node);
    }

    private static String identityId(JsonNode inp) {
        try {
            var seed = MAPPER.createObjectNode();
            seed.put("QuidID", inp.get("quidId").asText());
            seed.put("Name", inp.get("name").asText());
            seed.put("Creator", inp.get("creator").asText());
            seed.put("TrustDomain", inp.get("trustDomain").asText());
            seed.put("UpdateNonce", inp.get("updateNonce").asLong());
            seed.put("Timestamp", inp.get("timestamp").asLong());
            byte[] bytes = MAPPER.writeValueAsBytes(seed);
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return Hex.encode(md.digest(bytes));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] eventSignable(JsonNode inp) {
        var node = MAPPER.createObjectNode();
        node.put("id", inp.get("id").asText());
        node.put("type", "EVENT");
        node.put("trustDomain", inp.get("trustDomain").asText());
        node.put("timestamp", inp.get("timestamp").asLong());
        node.put("signature", "");
        node.put("publicKey", inp.get("publicKey").asText());
        node.put("subjectId", inp.get("subjectId").asText());
        node.put("subjectType", inp.get("subjectType").asText());
        node.put("sequence", inp.get("sequence").asLong());
        node.put("eventType", inp.get("eventType").asText());
        if (inp.hasNonNull("payload") && inp.get("payload").size() > 0) {
            node.set("payload", inp.get("payload"));
        }
        if (inp.hasNonNull("payloadCid") && !inp.get("payloadCid").asText().isEmpty()) {
            node.put("payloadCid", inp.get("payloadCid").asText());
        }
        if (inp.hasNonNull("previousEventId") && !inp.get("previousEventId").asText().isEmpty()) {
            node.put("previousEventId", inp.get("previousEventId").asText());
        }
        return nodeToBytes(node);
    }

    private static String eventId(JsonNode inp) {
        try {
            var seed = MAPPER.createObjectNode();
            seed.put("SubjectID", inp.get("subjectId").asText());
            seed.put("EventType", inp.get("eventType").asText());
            seed.put("Sequence", inp.get("sequence").asLong());
            seed.put("TrustDomain", inp.get("trustDomain").asText());
            seed.put("Timestamp", inp.get("timestamp").asLong());
            byte[] bytes = MAPPER.writeValueAsBytes(seed);
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return Hex.encode(md.digest(bytes));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] nodeToBytes(JsonNode node) {
        try {
            return MAPPER.writeValueAsBytes(node);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Put a float on an ObjectNode in Go-compatible form: integer-
     * valued floats serialize as integers, matching Go's
     * encoding/json behavior of eliding ".0" on whole-number
     * floats. Non-integer floats serialize as floats.
     */
    private static void putGoCompatFloat(
            com.fasterxml.jackson.databind.node.ObjectNode node,
            String key, double value) {
        if (Double.isFinite(value) && value == Math.floor(value) && Math.abs(value) < 1e15) {
            node.put(key, (long) value);
        } else {
            node.put(key, value);
        }
    }

    // -----------------------------------------------------------
    // Dynamic tests
    // -----------------------------------------------------------

    @TestFactory
    Stream<DynamicTest> trustVectors() throws Exception {
        JsonNode vf = loadVectorFile("trust-tx.json");
        assertEquals("TRUST", vf.get("tx_type").asText());
        Map<String, JsonNode> keys = loadKeys();
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : vf.get("cases")) {
            String name = c.get("name").asText();
            tests.add(DynamicTest.dynamicTest(name, () -> {
                JsonNode key = keys.get(c.get("signer_key_ref").asText());
                JsonNode inp = c.get("input");
                byte[] signable = trustSignable(inp);
                String id = trustId(inp);
                runCase(c, key, signable, id);
            }));
        }
        return tests.stream();
    }

    @TestFactory
    Stream<DynamicTest> identityVectors() throws Exception {
        JsonNode vf = loadVectorFile("identity-tx.json");
        assertEquals("IDENTITY", vf.get("tx_type").asText());
        Map<String, JsonNode> keys = loadKeys();
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : vf.get("cases")) {
            String name = c.get("name").asText();
            tests.add(DynamicTest.dynamicTest(name, () -> {
                JsonNode key = keys.get(c.get("signer_key_ref").asText());
                JsonNode inp = c.get("input");
                byte[] signable = identitySignable(inp);
                String id = identityId(inp);
                runCase(c, key, signable, id);
            }));
        }
        return tests.stream();
    }

    @TestFactory
    Stream<DynamicTest> eventVectors() throws Exception {
        JsonNode vf = loadVectorFile("event-tx.json");
        assertEquals("EVENT", vf.get("tx_type").asText());
        Map<String, JsonNode> keys = loadKeys();
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : vf.get("cases")) {
            String name = c.get("name").asText();
            tests.add(DynamicTest.dynamicTest(name, () -> {
                JsonNode key = keys.get(c.get("signer_key_ref").asText());
                JsonNode inp = c.get("input");
                byte[] signable = eventSignable(inp);
                String id = eventId(inp);
                runCase(c, key, signable, id);
            }));
        }
        return tests.stream();
    }

    @Test
    void sdkSignProducesIEEE1363() throws Exception {
        Quid q = Quid.generate();
        byte[] data = "test-data".getBytes(StandardCharsets.UTF_8);
        String sigHex = q.sign(data);
        assertEquals(128, sigHex.length(),
                "v1.0 mandates 64-byte IEEE-1363 (128 hex chars)");
        assertTrue(q.verify(data, sigHex), "SDK Sign/Verify round-trip failed");
    }
}

File: clients/rust/tests/vectors_test.rs
//! Rust SDK consumer for the v1.0 cross-SDK test vectors at
//! `docs/test-vectors/v1.0/`.
//!
//! Asserts the five conformance properties for each case in
//! `trust-tx.json`, `identity-tx.json`, `event-tx.json`, plus
//! the divergence probes that self-heal when the SDK converges.
//!
//! Run with: `cargo test --test vectors_test`.

use quidnug::wire::{IdentityTx, TrustTx};
use quidnug::Quid;
use serde::Deserialize;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
struct VectorFile {
    schema_version: String,
    tx_type: String,
    cases: Vec<VectorCase>,
}

#[derive(Debug, Deserialize)]
struct VectorCase {
    name: String,
    #[serde(default)]
    comments: String,
    signer_key_ref: String,
    input: serde_json::Value,
    expected: Expected,
}

#[derive(Debug, Deserialize)]
struct Expected {
    canonical_signable_bytes_hex: String,
    canonical_signable_bytes_utf8: String,
    sha256_of_canonical_hex: String,
    expected_id: String,
    reference_signature_hex: String,
    signature_length_bytes: usize,
}

#[derive(Debug, Deserialize)]
struct KeyFile {
    name: String,
    #[serde(default)]
    seed: String,
    private_scalar_hex: String,
    public_key_sec1_hex: String,
    quid_id: String,
}

fn vectors_root() -> PathBuf {
    // tests/ runs with the crate root as CWD, so go three up to
    // the repo root then into docs/test-vectors/v1.0.
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    p.pop(); // clients/
    p.pop(); // <repo>
    p.push("docs");
    p.push("test-vectors");
    p.push("v1.0");
    p
}

fn load_vector_file(name: &str) -> VectorFile {
    let path = vectors_root().join(name);
    let raw = fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("read {}: {}", path.display(), e));
    serde_json::from_str(&raw)
        .unwrap_or_else(|e| panic!("parse {}: {}", path.display(), e))
}

fn load_keys() -> std::collections::HashMap<String, KeyFile> {
    let dir = vectors_root().join("test-keys");
    let mut out = std::collections::HashMap::new();
    for entry in fs::read_dir(&dir).expect("read test-keys") {
        let entry = entry.expect("entry");
        if entry.path().extension().and_then(|s| s.to_str()) != Some("json") {
            continue;
        }
        let raw = fs::read_to_string(entry.path()).expect("read");
        let k: KeyFile = serde_json::from_str(&raw).expect("parse");
        out.insert(k.name.clone(), k);
    }
    out
}

/// Assert the five conformance properties for a single case.
fn run_case<F>(case: &VectorCase, keys: &std::collections::HashMap<String, KeyFile>, build_signable: F)
where
    F: FnOnce(&VectorCase, &KeyFile) -> (Vec<u8>, String),
{
    let key = keys
        .get(&case.signer_key_ref)
        .unwrap_or_else(|| panic!("no key ref {}", case.signer_key_ref));

    let (signable, derived_id) = build_signable(case, key);

    // Property 1: sha256 of canonical matches.
    use sha2::{Digest, Sha256};
    let sum = Sha256::digest(&signable);
    assert_eq!(
        hex::encode(sum),
        case.expected.sha256_of_canonical_hex,
        "{}: SHA-256 mismatch",
        case.name
    );

    // Property 2: canonical bytes match hex/utf8 in vector.
    assert_eq!(
        hex::encode(&signable),
        case.expected.canonical_signable_bytes_hex,
        "{}: hex canonical bytes mismatch",
        case.name
    );
    assert_eq!(
        std::str::from_utf8(&signable).expect("utf8"),
        case.expected.canonical_signable_bytes_utf8,
        "{}: utf8 canonical bytes mismatch",
        case.name
    );

    // Property 3: derived ID matches.
    assert_eq!(
        derived_id, case.expected.expected_id,
        "{}: ID derivation mismatch",
        case.name
    );

    // Property 4: reference signature verifies via Quid::verify
    // (public SDK API).
    let q_ro = Quid::from_public_hex(&key.public_key_sec1_hex)
        .unwrap_or_else(|e| panic!("from_public_hex: {e}"));
    assert!(
        q_ro.verify(&signable, &case.expected.reference_signature_hex),
        "{}: SDK Verify rejected reference signature",
        case.name
    );
    assert_eq!(
        case.expected.signature_length_bytes, 64,
        "{}: expected_sig_len != 64",
        case.name
    );

    // Property 5: tampered signature rejects.
    let mut tampered = hex::decode(&case.expected.reference_signature_hex).unwrap();
    tampered[5] ^= 0x01;
    assert!(
        !q_ro.verify(&signable, &hex::encode(&tampered)),
        "{}: tampered signature accepted",
        case.name
    );

    // Property 6: independent SDK sign-then-verify round-trip.
    let q_sign = Quid::from_private_hex_scalar(&key.private_scalar_hex)
        .unwrap_or_else(|e| panic!("from_private_hex_scalar: {e}"));
    let sdk_sig = q_sign.sign(&signable).expect("sign");
    assert_eq!(
        hex::decode(&sdk_sig).unwrap().len(),
        64,
        "{}: SDK produced non-64-byte signature",
        case.name
    );
    assert!(
        q_sign.verify(&signable, &sdk_sig),
        "{}: SDK sign-verify round-trip failed",
        case.name
    );

    // Quid ID derivation: sha256(pubkey)[:8] hex.
    let pub_bytes = hex::decode(&key.public_key_sec1_hex).unwrap();
    let quid_hash = Sha256::digest(&pub_bytes);
    let derived_quid = hex::encode(&quid_hash[..8]);
    assert_eq!(derived_quid, key.quid_id, "quid_id derivation mismatch");
}

#[test]
fn trust_vectors() {
    let vf = load_vector_file("trust-tx.json");
    assert_eq!(vf.schema_version, "1.0");
    assert_eq!(vf.tx_type, "TRUST");
    assert!(!vf.cases.is_empty());

    let keys = load_keys();

    for case in &vf.cases {
        run_case(case, &keys, |case, key| {
            let inp = &case.input;
            // Extract typed fields. Optional fields default to "".
            let description = inp.get("description").and_then(|v| v.as_str()).unwrap_or("");
            let valid_until = inp.get("validUntil").and_then(|v| v.as_i64()).unwrap_or(0);
            let mut tx = TrustTx {
                id: String::new(),
                tx_type: "TRUST",
                trust_domain: inp["trustDomain"].as_str().unwrap(),
                timestamp: inp["timestamp"].as_i64().unwrap(),
                signature: String::new(),
                public_key: &key.public_key_sec1_hex,
                truster: inp["truster"].as_str().unwrap(),
                trustee: inp["trustee"].as_str().unwrap(),
                trust_level: inp["trustLevel"].as_f64().unwrap(),
                nonce: inp["nonce"].as_i64().unwrap(),
                description,
                valid_until,
            };
            let id = tx.derive_id();
            tx.id = id.clone();
            let signable = serde_json::to_vec(&tx).expect("serialize");
            (signable, id)
        });
    }
}

#[test]
fn identity_vectors() {
    let vf = load_vector_file("identity-tx.json");
    assert_eq!(vf.tx_type, "IDENTITY");
    assert!(!vf.cases.is_empty());

    let keys = load_keys();

    for case in &vf.cases {
        run_case(case, &keys, |case, key| {
            let inp = &case.input;
            let description = inp.get("description").and_then(|v| v.as_str()).unwrap_or("");
            let home_domain = inp.get("homeDomain").and_then(|v| v.as_str()).unwrap_or("");
            let attributes = inp.get("attributes").cloned();
            let mut tx = IdentityTx {
                id: String::new(),
                tx_type: "IDENTITY",
                trust_domain: inp["trustDomain"].as_str().unwrap(),
                timestamp: inp["timestamp"].as_i64().unwrap(),
                signature: String::new(),
                public_key: &key.public_key_sec1_hex,
                quid_id: inp["quidId"].as_str().unwrap(),
                name: inp["name"].as_str().unwrap(),
                description,
                attributes,
                creator: inp["creator"].as_str().unwrap(),
                update_nonce: inp["updateNonce"].as_i64().unwrap(),
                home_domain,
            };
            let id = tx.derive_id();
            tx.id = id.clone();
            let signable = serde_json::to_vec(&tx).expect("serialize");
            (signable, id)
        });
    }
}

#[test]
fn sdk_sign_produces_ieee1363() {
    // Regression probe: Quid::sign must produce 64-byte IEEE-1363.
    let q = Quid::generate();
    let sig_hex = q.sign(b"test-data").expect("sign");
    let sig_bytes = hex::decode(&sig_hex).expect("hex");
    assert_eq!(sig_bytes.len(), 64,
        "v1.0 requires 64-byte IEEE-1363; got {} bytes", sig_bytes.len());
    assert!(q.verify(b"test-data", &sig_hex),
        "SDK Sign/Verify round-trip failed");
}
